### PR TITLE
Create skeleton code for degraded mode procedure

### PR DESCRIPTION
**Pull request by sawsa307**
Create skeleton code for degraded mode procedure. Implementation for filling missing information and filter invalid pod will come in separate PRs.

### CODE REVIEW

File: pkg/neg/syncers/utils.go [hunk 1]
```diff
@@ -296,6 +297,74 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
 	return zoneNetworkEndpointMap, networkEndpointPodMap, dupCount, nil
 }
 
+func toZoneNetworkEndpointMapDegradedMode(eds []negtypes.EndpointsData, zoneGetter negtypes.ZoneGetter, podLister cache.Indexer, servicePortName string, networkEndpointType negtypes.NetworkEndpointType) (map[string]negtypes.NetworkEndpointSet, negtypes.EndpointPodMap, int, error) {
+	targetMap := map[string]negtypes.NetworkEndpointSet{}
+	endpointPodMap := negtypes.EndpointPodMap{}
+	for _, ed := range eds {
+		// skip custom endpoint slice so it won't block the neg api call from succeeding
+		if val, ok := ed.Meta.Labels[discovery.LabelManagedBy]; !ok || val != "endpointslice-controller.k8s.io" {
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:305]: We should still validate custom endpoints. We should add them if they are valid pods.

Comment (swetharepakula) [reply pkg/neg/syncers/utils.go:305]: this also means we won't be adding the customer endpoints to the NEG?

Comment (sawsa307) [reply pkg/neg/syncers/utils.go]: Removed this part. Thanks!

File: pkg/neg/syncers/utils_test.go [hunk 2]
```diff
@@ -1190,6 +1192,298 @@ func TestNEGRecreate(t *testing.T) {
...
+				networkEndpointFromEncodedEndpoint("10.100.1.4||||80"): types.NamespacedName{Namespace: testNamespace, Name: "pod6"},
+			},
+			networkEndpointType: negtypes.NonGCPPrivateEndpointType,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			testEndpointSlices := getTestEndpointSlices(testService, testNamespace)
+
+			testEndpointSlices[2].ObjectMeta.Labels[discovery.LabelManagedBy] = tc.managedByLabel
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:1300]: add this as a testcase instead of modifying the EPS after the fact.

File: pkg/neg/syncers/utils.go [hunk 3]
```diff
@@ -296,6 +296,76 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+		}
+		endpointPodMap[networkEndpoint] = types.NamespacedName{Namespace: ep.TargetRef.Namespace, Name: ep.TargetRef.Name}
+	}
+	return dupCount
+}
+
+// validatePod checks if this pod is a valid pod resource
+func validatePod(pod *apiv1.Pod) bool {
+	return true
+}
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:367]: Add this check in only when it is ready.

File: pkg/neg/syncers/utils.go [hunk 4]
```diff
@@ -296,6 +296,76 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+		}
+		nodeName := pod.Spec.NodeName
+		zone, err := zoneGetter.GetZoneForNode(nodeName)
+		if err != nil {
+			continue
+		}
+
+		if endpointType == negtypes.NonGCPPrivateEndpointType {
+			// Non-GCP network endpoints don't have associated nodes.
+			nodeName = ""
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:345]: What do these endpoints look like? Will there be a nodename and a pod?

File: pkg/neg/syncers/utils.go (hunk 3 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils.go:367]: Removed. Thanks!

File: pkg/neg/syncers/utils_test.go (hunk 2 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils_test.go:1300]: Updated. Removing this since the part that filter custom endpoint slice is deleted https://github.com/kubernetes/ingress-gce/pull/1952#discussion_r1127198331.

File: pkg/neg/syncers/utils.go (hunk 4 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils.go:345]: These endpoints will have IP address must belong to a VM, and associated port if default-port is not specify.
This part is mirroring the behavior for https://github.com/kubernetes/ingress-gce/blob/master/pkg/neg/syncers/utils.go#L275-L278.

File: pkg/neg/syncers/utils.go [hunk 5]
```diff
@@ -296,6 +296,68 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+				matchPort = strconv.Itoa(int(port.Port))
+				break
+			}
+		}
+		if len(matchPort) == 0 {
+			continue
+		}
+		for _, endpointAddress := range ed.Addresses {
+			if endpointAddress.TargetRef == nil {
+				continue
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:316]: add a log to indicate no matching targetRef

File: pkg/neg/syncers/utils.go [hunk 6]
```diff
@@ -296,6 +296,68 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+}
+
+// validateAndAddEndpoints fills in missing information and creates network endpoint for each endpoint addresss
+func ValidateAndAddEndpoints(ep negtypes.AddressData, zoneGetter negtypes.ZoneGetter, podLister cache.Indexer, matchPort string, endpointType negtypes.NetworkEndpointType, targetMap map[string]negtypes.NetworkEndpointSet, endpointPodMap negtypes.EndpointPodMap) int {
+	var dupCount int
+	for _, address := range ep.Addresses {
+		key := fmt.Sprintf("%s/%s", ep.TargetRef.Namespace, ep.TargetRef.Name)
+		obj, exists, err := podLister.GetByKey(key)
+		if err != nil || !exists {
+			continue
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:331]: log here that we are skipping this endpoint.

File: pkg/neg/syncers/utils.go [hunk 7]
```diff
@@ -296,6 +296,68 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+		}
+		pod, ok := obj.(*apiv1.Pod)
+		if !ok {
+			continue
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:335]: log that we are skipping this endpoint

File: pkg/neg/syncers/utils.go [hunk 8]
```diff
@@ -296,6 +296,68 @@ func toZoneNetworkEndpointMap(eds []negtypes.EndpointsData, zoneGetter negtypes.
...
+		}
+		nodeName := pod.Spec.NodeName
+		zone, err := zoneGetter.GetZoneForNode(nodeName)
+		if err != nil {
+			continue
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils.go:340]: log that we are skipping this endpoint

File: pkg/neg/syncers/utils_test.go [hunk 9]
```diff
@@ -1190,6 +1192,290 @@ func TestNEGRecreate(t *testing.T) {
 	}
 }
 
+func TestToZoneNetworkEndpointMapDegradedMode(t *testing.T) {
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:1195]: can you add negative test cases as well to this.

File: pkg/neg/syncers/utils_test.go [hunk 10]
```diff
@@ -1250,6 +1536,7 @@ func getTestEndpointSlices(name, namespace string) []*discovery.EndpointSlice {
 	notReady := false
 	emptyNamedPort := ""
 	testNamedPort := testNamedPort
+	managedByController := "endpointslice-controller.k8s.io"
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:1539]: why is this change necessary?

**Review by swetharepakula — COMMENTED** [review]
Last remaining comments are about logging and negative test cases. Otherwise the functionality looks good to me.

File: pkg/neg/syncers/utils.go (hunk 5 above)
Comment (sawsa307) [reply pkg/neg/syncers/utils.go:316]: Added. Thanks!

File: pkg/neg/syncers/utils_test.go (hunk 9 above)
Comment (swetharepakula) [reply pkg/neg/syncers/utils_test.go:1195]: discussed offline, the e2e will be tested separately, and negative cases for now will be covered by the ValidateAndAddEndpoints

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve
